### ISSUE TITLE
Connection to database is returning 1 record less than in Studio

### ISSUE THREAD

**Issue by ShakAttacked** [orientechnologies/orientdb#8034]
## Actual behavior  
There are 192 records according to Studio but the connection returns 191, whether I select * or I count(*), it's always 1 less.  The missing record was created by a previous request that was successful and committed. This code is not new and worked in 2.2.

Here is what's mind boggling issue, at first I thought it's the connection pool, so I'm opening a new connection for every request and it hasn't changed anything. I still get 1 less record.  So I modified the code to open a new connection right after the failed request. the  new connection gets the proper count!!!!! so I had to open 2 connections, the first fails, the 2nd works.  

My uneducated analysis is that even though I'm opening a new connection on every request I'm getting some cached/corrupted connection. My tests are running in series, so it's very likely that it's reusing the same resources from the server??

I've tried to reproduce this in a small example but was not able. This requires someone with very good knowledge of the database code to take a guess on why would a newly opened connection return less records than Studio would?

we have an opportunity here because it's repeatable.  

Update: Even if I take my application down and up, the first connection would always return one less record.  the problem is at the server side where that first connection is returning the wrong data.

Safwan

**Comment by tglman** [comment]
Hi @ShakAttacked,

This is quite strange, first thing try to update to the 3.0.0 Final and see if you can still reproduce this problem, after that can you let us know if you are running in distributed (you run the server with "./dserver.sh") or standalone ("./server.sh"), and it would be really cool if you can provide a test case to reproduce that.

Regards

**Comment by ShakAttacked** [comment]
Hi,

I'll do that tonight. I can tell you already that it's with ./server.sh.  I have spent more than 12 hours to try to eliminate that the issue is coming from my code. and then tried to reproduce it with a simple scenario but I am not able to.  There is a lot happening such as dealing with arrays, hashmaps, multiple updates to the same vertex, edges, all in one action.   I'll give the latest snapshot a try and update you.

If it's still happening, I can easily clear the database and run that test alone we can turn on additional logs (if you tell me how and what you need turned on) to see.

Safwan

**Comment by ShakAttacked** [comment]
Hi,

I managed to test right away.  I don't have good news.  I've even disabled Hooks, used the default database config instead of my own.  As you can see the output is different on different connections.

old connection:{"count(*)": 9} 
new connection:{"count(*)": 10} 

I think I have one more idea to try. leave it with me for now.

Safwan

**Comment by laa** [comment]
Hi, @ShakAttacked was the issue resolved?

**Comment by ShakAttacked** [comment]
Hi @laa,

yes. it was on our side after all.  it was 2 separate problems.  A change in approach moving from 2.2 batch scripts to using vertexes in 3.0 and a mistake by a developer.  I will explain it here for others who might have the same problem.

1- As we are moving to 3.0, we have stopped using SQL batch scripts.  In the old code, a delete of a record in our code was an instruction to be executed when the batch script is committed.  So a select for that record would of course make it available until the transaction is committed.  However, when deleting a vertex in 3.0 (maybe earlier) a fetch of edges will exclude deleted vertex.  

2- The reason that I couldn't see this is that the programmer who wrote the code was deleting the record twice from 2 different places.  One in the base code which would delete the vertex (that's the proper place) and then he had another piece of code that would later attempt to fetch the record to get its ID and then delete it.  that's when we saw one less record.  By deleting his 2nd code the problem went away.

That's why we would see less records.  It's not that there are more records in the database.  There is a deleted vertex in the current transactions which is uncommitted and a fetch does not see it.  Was it like that in 2.2 as well?  

Sorry if I made you waste time on this.  but now, if someone else brings it up, it might help them figure it out.

Thanks,
Safwan
